### PR TITLE
Feature/glpi database check

### PR DESCRIPTION
**Pull request by cadegenn**
Hi,

This is a pull-request to enhance php bin/console glpi:database:check output. From a fresh GLPI install, it outputs lots of false positives.

It replaces PR #8572 following instructions provided by @cedric-anne

Hope I did it right.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | 
| Fixed tickets | #8571

### CODE REVIEW

File: install/mysql/glpi-empty.sql [hunk 1]
```diff
@@ -1249,7 +1241,7 @@ CREATE TABLE `glpi_impactitems` (
 DROP TABLE IF EXISTS `glpi_impactcontexts`;
 CREATE TABLE `glpi_impactcontexts` (
   `id` INT NOT NULL AUTO_INCREMENT,
-  `positions` TEXT NOT NULL,
```
Comment (trasher) [inline install/mysql/glpi-empty.sql:1252]: I do not understand those ones. All other types seems to be uppercase.

**Comment by trasher** [comment]
@cedric-anne actions matrix should be temporary updated  for this one so we can see all fails; am I right?

File: install/mysql/glpi-empty.sql [hunk 2]
```diff
@@ -21,7 +21,7 @@
 --
 -- GLPI is distributed in the hope that it will be useful,
 -- but WITHOUT ANY WARRANTY; without even the implied warranty of
--- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
```
Comment (trasher) [inline install/mysql/glpi-empty.sql:24]: This one breaks the lint process
```suggestion
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
```

File: install/mysql/glpi-empty.sql (hunk 1 above)
Comment (cadegenn) [reply install/mysql/glpi-empty.sql:1252]: Types are mixed upper and lowercase. Changing this case was a test for my regex. I didn't revert it afterward sorry.

File: install/mysql/glpi-empty.sql (hunk 2 above)
Comment (cadegenn) [reply install/mysql/glpi-empty.sql:24]: yes I saw that too late. I revert it back

**Comment by cedric-anne** [comment]
> @cedric-anne actions matrix should be temporary updated for this one so we can see all fails; am I right?

Yes, you shopuld apply this patch:
```diff
diff --git a/.github/workflows/ci.yml b/.github/workflows/ci.yml
index 9bd4f2146a..35edb56ff1 100644
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -138,10 +138,15 @@ jobs:
           - {php-version: "7.3", db-image: "mariadb:10.5", always: false}
           - {php-version: "7.4", db-image: "mariadb:10.5", always: false}
           # test other DB servers/versions
-          - {php-version: "8.0", db-image: "mariadb:10.1", always: false}
-          - {php-version: "8.0", db-image: "mysql:5.6", always: false}
-          - {php-version: "8.0", db-image: "percona:5.6", always: false}
-          - {php-version: "8.0", db-image: "percona:8.0", always: false}
+          - {php-version: "8.0", db-image: "mariadb:10.1", always: true}
+          - {php-version: "8.0", db-image: "mysql:5.6", always: true}
+          - {php-version: "8.0", db-image: "percona:5.6", always: true}
+          - {php-version: "8.0", db-image: "percona:8.0", always: true}
+          # temporarly test on all DB versions
+          - {php-version: "8.0", db-image: "mariadb:10.2", always: true}
+          - {php-version: "8.0", db-image: "mariadb:10.3", always: true}
+          - {php-version: "8.0", db-image: "mariadb:10.4", always: true}
+          - {php-version: "8.0", db-image: "mysql:5.7", always: true}
     env:
       # Skip jobs that should not be always run on pull requests or on push on tier repository (to limit workers usage).
       # No jobs will be skipped on nightly build, manual dispatch or push on main branches (master and */bugfixes) or tags.
```

File: install/mysql/glpi-empty.sql (hunk 1 above)
Comment (cadegenn) [reply install/mysql/glpi-empty.sql:1252]: If my regex is right : 

```
grep "\`[[:alnum:]_]\+\` [a-z]\+" install/mysql/glpi-empty.sql | wc -l
grep "\`[[:alnum:]_]\+\` [A-Z]\+" install/mysql/glpi-empty.sql | wc -l
```
- nb of lowercase types : 3613
- nb of uppercase types : 66

:open_mouth:

Comment (cedric-anne) [reply install/mysql/glpi-empty.sql:1252]: IMHO, as `SHOW CREATE TABLE` seems to always return lowercased types, changing these 66 occurences of uppercased types is an acceptable change (and will not be so hard to validate).

Comment (cadegenn) [reply install/mysql/glpi-empty.sql:1252]: done in 89e501aa2385b85e942432c72c70df0da1cabd25

**Comment by trasher** [comment]
@cadegenn as you may have noticed, some of the github actions are failing on the db:check command. They all must be fixed for the change to be merged :)

**Comment by cadegenn** [comment]
Yes I saw that :relaxed: ... I was waiting for futher instruction, without asking for it, sorry :confused:.
I think I could add missing default values in `glpi-empty.sql` ?
Fortunately, the result given by all database engines seems to be the same.

I don't know what to do with this one : 
```
-  foreign key (agenttypes_id) references glpi_agenttypes (id) on delete cascade on update cascade
+  constraint `glpi_agents_ibfk_1` foreign key (`agenttypes_id`) references `glpi_agenttypes` (`id`) on delete cascade on update cascade
```

**Comment by trasher** [comment]
> I think I could add missing default values in `glpi-empty.sql` ?

Yes, you can now do the required changes to get the script fixed.

> I don't know what to do with this one :
> 
> ```
> -  foreign key (agenttypes_id) references glpi_agenttypes (id) on delete cascade on update cascade
> +  constraint `glpi_agents_ibfk_1` foreign key (`agenttypes_id`) references `glpi_agenttypes` (`id`) on delete cascade on update cascade
> ```

Hum... I guess the `constraint` part should just be ignored, if possible; it's not the important part of the information :)

**Comment by cadegenn** [comment]
ok I'll do that.
thanks

**Comment by cedric-anne** [comment]
> > I don't know what to do with this one :
> > ```
> > -  foreign key (agenttypes_id) references glpi_agenttypes (id) on delete cascade on update cascade
> > +  constraint `glpi_agents_ibfk_1` foreign key (`agenttypes_id`) references `glpi_agenttypes` (`id`) on delete cascade on update cascade
> > ```
> 
> Hum... I guess the `constraint` part should just be ignored, if possible; it's not the important part of the information :)

We specify names for other keys, so we may also change the line to this:
```
constraint `glpi_agents_fk_agenttypes_id` foreign key (`agenttypes_id`) references `glpi_agenttypes` (`id`) on delete cascade on update cascade
```

**Comment by trasher** [comment]
OK for me; this one is new in master, its does not require a migration.

**Comment by cedric-anne** [comment]
> > > I don't know what to do with this one :
> > > ```
> > > -  foreign key (agenttypes_id) references glpi_agenttypes (id) on delete cascade on update cascade
> > > +  constraint `glpi_agents_ibfk_1` foreign key (`agenttypes_id`) references `glpi_agenttypes` (`id`) on delete cascade on update cascade
> > > ```
> > 
> > 
> > Hum... I guess the `constraint` part should just be ignored, if possible; it's not the important part of the information :)
> 
> We specify names for other keys, so we may also change the line to this:
> 
> ```
> constraint `glpi_agents_fk_agenttypes_id` foreign key (`agenttypes_id`) references `glpi_agenttypes` (`id`) on delete cascade on update cascade
> ```

It does not work, maybe it should be handled by removing the ``CONSTAINT\s+`[^`]+`\s+`` part of the line.

**Comment by cadegenn** [comment]
hmm strange.
```
-  constraint `glpi_agents_fk_agenttypes_id` foreign key (`agenttypes_id`) references `glpi_agenttypes` (`id`) on delete cascade on update cascade
+  constraint `glpi_agents_ibfk_1` foreign key (`agenttypes_id`) references `glpi_agenttypes` (`id`) on delete cascade on update cascade
```
on my test machine, if I create the database from scratch, mysql return the right constraint name `glpi_agents_fk_agenttypes_id`, not (what I think is) a generated one.

**Comment by cedric-anne** [comment]
> hmm strange.
> 
> ```
> -  constraint `glpi_agents_fk_agenttypes_id` foreign key (`agenttypes_id`) references `glpi_agenttypes` (`id`) on delete cascade on update cascade
> +  constraint `glpi_agents_ibfk_1` foreign key (`agenttypes_id`) references `glpi_agenttypes` (`id`) on delete cascade on update cascade
> ```
> 
> on my test machine, if I create the database from scratch, mysql return the right constraint name `glpi_agents_fk_agenttypes_id`, not (what I think is) a generated one.

Indeed, problem is now on update.
I think you just have to fix it here : https://github.com/glpi-project/glpi/blob/master/install/update_95_xx/native_inventory.php#L82 .

**Comment by cadegenn** [comment]
I don't understand the purpose of this file `native_inventory.php`. Do I need to update it to reflect all changes from `glpi-empty.sql` ?

**Comment by cadegenn** [comment]
What I now understand is that I have to create/generate/update the migration code to include changes of `glpi-empty.sql`. I'll investigate on how to do that.

**Comment by cedric-anne** [comment]
> What I now understand is that I have to create/generate/update the migration code to include changes of `glpi-empty.sql`. I'll investigate on how to do that.

No, you should not change migrations other than thoose already done.

I guess your last errors you have are related to the `tests/glpi-*-empty.sql` files. If you can handle this difference without touching them, it would be an acceptable solution.

**Comment by cadegenn** [comment]
I though about adding `ALTER TABLE`s in `update_95_xx.php` to update existing DB to reach target `glpi-empty.sql` state.
Am I wrong ?

**Comment by cedric-anne** [comment]
> I though about adding `ALTER TABLE` in `update_95_xx.php` to update existing DB to reach target `glpi-empty.sql` state.
> Am I wrong ?

It may be a good solution.

**Comment by cadegenn** [comment]
damned ! I though fixing an indentation was harmless ...

**Comment by cedric-anne** [comment]
> damned ! I though fixing an indentation was harmless ...

Failure does not seems to be related to your pull request. Try to rebase on master branch to see if problem is fixed.

**Comment by cadegenn** [comment]
thank you... hope I did it right

**Comment by cedric-anne** [comment]
> thank you... hope I did it right

Not sure, seems that you pushed some extra commit. If you agree with that, I can do it on my side and push the result on your branch (I should have rights to do this, by default).

**Comment by cadegenn** [comment]
I agree...

**Comment by cedric-anne** [comment]
> I agree...

I've rebase. Problem exists on current master, so it is not related to your changes. I will try to fix it.

File: inc/dbmysql.class.php [hunk 3]
```diff
@@ -1364,6 +1364,24 @@ function($idx) { return rtrim($idx, ','); },
...
+        "/ (TEXT)\s+NULL/i",
+        " $1",
+        $structure
+      );
+      $structure = preg_replace(
+        "/ (INT)\s+NULL/i",
+        " $1",
+        $structure
+      );
+
```
Comment (cedric-anne) [inline inc/dbmysql.class.php:1384]: There are only a few occurences of this in the installation file. If all supported DB servers/versions are acting the same, I would prefer to directly remove this extra NULL in the `glpi-empty.sql` file.
MySQL doc says: `If neither NULL nor NOT NULL is specified, the column is treated as though NULL had been specified.`

Comment (cadegenn) [reply inc/dbmysql.class.php:1384]: I agree with that. It would be less code to maintain.
I'll do it.

**Review by cedric-anne — COMMENTED** [review]
@cadegenn 

:+1: 

This PR is OK for me. Failing tests are not due to your changes. I will investigate on it and eventally, push some commits to your branch.

Please do not push new commits to prevent conflicts.

**Comment by cadegenn** [comment]
ok roger that :+1:

**Comment by cedric-anne** [comment]
I have fixed test suite on master, rebased your branch on master, and add a commit to remove the `ON UPDATE CURRENT_TIMESTAMP` addition you made on some date fields. Indeed, using the current timestamp on creation is correct, but on these tables, this field should contains the creation date, and should not be updated automatically on update.

Tests should pass now.

**Review by cedric-anne — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -138,10 +138,15 @@ jobs:
           - {php-version: "7.3", db-image: "mariadb:10.5", always: false}
           - {php-version: "7.4", db-image: "mariadb:10.5", always: false}
           # test other DB servers/versions
-          - {php-version: "8.0", db-image: "mariadb:10.1", always: false}
```
Comment (trasher) [inline .github/workflows/ci.yml:141]: Please remove ;)

**Review by trasher — APPROVED** [review]
OK for me, thanks :)

**Pull request merged**